### ISSUE TITLE
Improvements to FSDP debugability

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#98484]
### 🚀 The feature, motivation and pitch

There are a couple pain points which make FSDP harder to debug: 
- In some cases, post backward hooks don't fire resulting in layers not getting gradients, resulting in training convergence issues. We could add logging in a debug mode for this, but need to think of a more comprehensive solution to identify this issue. 
- On lazy init, we should iterate through the original params and check that shared params are in the same FSDP instance.
